chat/src/main.rs

```diff
@@ -0,0 +1,132 @@
+        )
+        .to_string()
+});
+
+static ICON: Lazy<Icon> = Lazy::new(|| {
+    let bytes: Vec<u8> = include_bytes!("../icon.png").to_vec();
+    let imagebuffer = image::load_from_memory_with_format(&bytes, ImageFormat::Png)
+        .unwrap()
+        .into_rgba8();
+    let (icon_width, icon_height) = imagebuffer.dimensions();
+    let icon_rgba = imagebuffer.into_raw();
+    Icon::from_rgba(icon_rgba, icon_width, icon_height).unwrap()
+});
+
+fn main() -> wry::Result<()> {
+    // println!("{}", serde_json::to_string_pretty(&Message::Focus).unwrap());
+    let event_loop = EventLoop::with_user_event();
+    let window = WindowBuilder::new()
+        .with_title(WINDOW_TITLE)
+        .with_window_icon(Some(ICON.clone()))
+        .build(&event_loop)?;
+    let webview = WebViewBuilder::new(window)?
+        .with_custom_protocol("wry".into(), |request| {
+            let path = request.uri().path();
+            // Read the file content from file path
+            let content = if path == "/" {
+                INDEX_HTML.as_bytes().into()
+            } else {
+                read(canonicalize(
+                    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(&path[1..]),
+                )?)?
+                .into()
+            };
+
+            let mimetype = if path.ends_with(".html") || path == "/" {
+                "text/html"
+            } else if path.ends_with(".js") {
+                "text/javascript"
+            } else {
+                unimplemented!();
+            };
+
+            wry::http::Response::builder()
+                .header(wry::http::header::CONTENT_TYPE, mimetype)
+                .body(content)
+                .map_err(Into::into)
+        })
+        .with_ipc_handler(move |_window: &Window, req: String| {
+            let mut lock = io::stdout().lock();
+            let _ = writeln!(lock, "{req}");
+        })
+        .with_url(BASE_URL)?
+        .build()?;
+
+    let proxy = event_loop.create_proxy();
+    thread::spawn(move || loop {
+        let mut buffer = String::new();
+        io::stdin().read_line(&mut buffer).unwrap();
+        let _ = proxy.send_event(buffer);
+    });
+
+    event_loop.run(move |event, _, control_flow| {
+        *control_flow = ControlFlow::Wait;
+
+        match event {
+            Event::NewEvents(StartCause::Init) => (),
+            Event::WindowEvent {
+                event: WindowEvent::CloseRequested,
+                ..
+            } => *control_flow = ControlFlow::Exit,
+            Event::UserEvent(message) => match serde_json::from_str::<Message>(&message) {
+                Ok(Message::Focus) => {
+                    webview.window().set_focus();
+                    if env::consts::OS == "linux" {
+                        let _ = std::process::Command::new("wmctrl")
+                            .args(["-a", WINDOW_TITLE])
+                            .output();
+                    }
+                }
```

Comment:
This *seems* to work without wmctrl, but I just want to make sure to bring attention to this (optional?) dependency on wmctrl. When I was messing around with this, it seems that it just sets `urgent` on the window when not installed. `webview.window().set_focus()`(?), and nothing else?

Reply:
From reading the docs, it seems that this returns a `Result`, which as long as it isn't attempted to be unwrapped, then it shouldn't panic, which explains the current behavior (which is desired behavior).

Reply:
`set_focus` unfortunately is not respected by all the window managers (my kde doesn't  seem to respect it) so this is why this fallback. This is not ideal to sat the least but this is what I've found so far